mat64/shadow.go

```diff
@@ -0,0 +1,201 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat64
+
+import "github.com/gonum/blas/blas64"
+
+const (
+	// regionOverlap is the panic string used for the general case
+	// of a matrix region overlap between a source and destination.
+	regionOverlap = "mat64: bad region: overlap"
+
+	// regionIdentity is the panic string used for the specific
+	// case of complete agreement between a source and a destination.
+	regionIdentity = "mat64: bad region: identical"
+
+	// mismatchedStrides is the panic string used for overlapping
+	// data slices with differing strides.
+	mismatchedStrides = "mat64: bad region: different strides"
+)
+
+// checkOverlap returns false if the receiver does not overlap data elements
+// referenced by the parameter and panics otherwise.
+//
+// checkOverlap methods return a boolean to allow the check call to be added to a
+// boolean expression, making use of short-circuit operators.
+
+func (m *Dense) checkOverlap(a blas64.General) bool {
```

Comment:
It seems like we could make it simple with something like

````
func (m *Dense) checkOverlap(a Matrix) bool {
    if m == a {
         return false
    }
     // Type switch to extract the data. If a is none of the Raw types, return false since we can't check.
    // If offset is in the right range, and the types are different, panic
    // If a is a RawMatrixer, do the fancier checks
}
````
This way, we can do something like
````
func (m *Dense) Method(a Matrix) {
    // Check size
    aMat, _ := untranspose(a)
    m.checkOverlap(aMat)
}
````
Thus we can check all of the necessary overlap at the top of the function once. We could even remove the boolean.